src/util/JoinAlgorithms/JoinAlgorithms.h

```diff
@@ -912,28 +996,64 @@ struct BlockZipperJoinImpl {
         getFirstBlock(currentBlocksRight, currentEl);
 
     compatibleRowAction_.setInput(fullBlockLeft.get(), fullBlockRight.get());
-    auto addRowIndex = [begL = fullBlockLeft.get().begin(),
-                        begR = fullBlockRight.get().begin(),
-                        this](auto itFromL, auto itFromR) {
+    auto begL = fullBlockLeft.get().begin();
+    auto begR = fullBlockRight.get().begin();
+    auto addRowIndex = [&begL, &begR, this](auto itFromL, auto itFromR) {
       compatibleRowAction_.addRow(itFromL - begL, itFromR - begR);
     };
 
     auto addNotFoundRowIndex = [&]() {
       if constexpr (DoOptionalJoin) {
-        return [begL = fullBlockLeft.get().begin(), this](auto itFromL) {
+        return [this, begL = fullBlockLeft.get().begin(),
+                &subrangeRight](auto itFromL) {
+          if constexpr (potentiallyHasUndef) {
+            AD_CORRECTNESS_CHECK(!isUndefined_(subrangeRight.front()) &&
+                                 rightSide_.undefBlocks_.empty());
+          } else {
+            (void)subrangeRight;
+          }
           compatibleRowAction_.addOptionalRow(itFromL - begL);
         };
 
       } else {
         return ad_utility::noop;
       }
     }();
-    [[maybe_unused]] auto res = zipperJoinWithUndef(
-        std::ranges::subrange{subrangeLeft.begin(), currentElItL},
-        std::ranges::subrange{subrangeRight.begin(), currentElItR}, lessThan_,
-        addRowIndex, noop, noop, addNotFoundRowIndex);
+    if constexpr (potentiallyHasUndef) {
```

Comment:
Can you do a (then non-constexpr, but optimized by the compiler)
` if (potentiallyHasUndef && actuallyHasUndef)` for performance reasons in the case where we can't prove in advance that there are no undef values, but that in fact is the case?
Or is the plan to preconsume the generators to always know for sure in advance whether we have undefined values?

Reply:
Tried it, but then the compilation seems to no longer work. Haven't been able to figure out the exact cause so far.